platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ExportMigrationContextImpl.java

```diff
@@ -0,0 +1,278 @@
+                    name));
+            return Optional.empty();
+        }
+        final ExportMigrationSystemPropertyReferencedEntryImpl sprop =
+                new ExportMigrationSystemPropertyReferencedEntryImpl(this, name, val);
+
+        systemProperties.put(name, sprop);
+        return Optional.of(sprop);
+    }
+
+    @Override
+    public ExportMigrationEntry getEntry(Path path) {
+        final ExportMigrationEntryImpl e = new ExportMigrationEntryImpl(this, path);
+
+        return entries.computeIfAbsent(e.getPath(), p -> e);
+    }
+
+    @Override
+    public Stream<ExportMigrationEntry> entries(Path path) {
+        final ExportMigrationEntryImpl entry = new ExportMigrationEntryImpl(this, path);
+
+        if (!isDirectory(entry)) {
+            return Stream.empty();
+        }
+        return FileUtils.listFiles(entry.getFile(),
+                TrueFileFilter.INSTANCE,
+                TrueFileFilter.INSTANCE)
+                .stream()
+                .map(File::toPath)
+                .map(this::getEntry);
+    }
+
+    @Override
+    public Stream<ExportMigrationEntry> entries(Path path, PathMatcher filter) {
+        final ExportMigrationEntryImpl entry = new ExportMigrationEntryImpl(this, path);
+
+        Validate.notNull(filter, "invalid null filter");
+        if (!isDirectory(entry)) {
+            return Stream.empty();
+        }
+        return FileUtils.listFiles(entry.getFile(),
+                TrueFileFilter.INSTANCE,
+                TrueFileFilter.INSTANCE)
+                .stream()
+                .map(File::toPath)
+                .map(p -> new ExportMigrationEntryImpl(this, p))
+                .filter(e -> filter.matches(e.getPath()))
+                .map(e -> entries.computeIfAbsent(e.getPath(), p -> e));
+    }
+
+    @Override
+    public void close() throws IOException {
+        final OutputStream oos = this.currentOutputStream;
+
+        if (oos != null) {
+            this.currentOutputStream = null;
+            oos.close();
+        }
+    }
+
+    /**
+     * Performs an export using the context's migratable.
+     *
+     * @return metadata to export for the corresponding migratable keyed by the migratable's id
+     * @throws org.codice.ddf.migration.MigrationException to stop the export operation
+     */
+    // PMD.DefaultPackage - designed to be called from ExportMigrationManagerImpl within this package
+    @SuppressWarnings("PMD.DefaultPackage")
+    Map<String, Map<String, Object>> doExport() {
+        LOGGER.debug("Exporting [{}] with version [{}]...",
+                id,
+                getVersion().orElse(null)); // version will never be empty
+        Stopwatch stopwatch = null;
+
+        if (LOGGER.isDebugEnabled()) {
+            stopwatch = Stopwatch.createStarted();
+        }
+        migratable.doExport(this);
+        if (LOGGER.isDebugEnabled() && (stopwatch != null)) {
```

Comment:
❓ What should we do if a migratable hangs during the export? Should we thread this out and stop the thread if it takes too long?

Reply:
That is the original code. Plus it was originally documented that export was single threaded. I'd be willing to re-design this but that might be more work. Is it worth it?